### ISSUE TITLE
Support for more databases

### ISSUE THREAD

**Issue by fforootd** [zitadel/zitadel#3598]
Based on a discussion on [hackernews] (https://news.ycombinator.com/item?id=31266460)it might make sense for us to support multiple storages.
Even though CockroachDB will be the primary one

**Acceptance criteria**
- [ ] Support Postgresql
- [ ] Discuss what more we can support

**Comment by fforootd** [comment]
I like to hear your thoughts on this @adlerhurst

**Comment by adlerhurst** [comment]
IMO before we begin we should discuss how we will test other databases. They might need different indexes. We can assume that we have to create the same indexes and index types as for crdb. 

To not break the configuration files structure for v2 we have to add a database type, this must be done before the release.

**Comment by adlerhurst** [comment]
Added task for database type: #3600

**Comment by fforootd** [comment]
I did some research around the offerings from the hyperscalers. It turns out that >80% of them has some form of pure Postgresql offering. Based on this I think we should allow and test for plain Postgresql as well besides cockroachDB.

What I currently would ignore are their proprietary offerings e.g CosmosDB, Amazon RDS, Google Spanner. Even though some of them provide Postgresql as dialect

**Comment by Sytten** [comment]
We are interested in using zitadel and being able to use postgres is the largest hurdle for adoption on our side.

**Comment by fforootd** [comment]
@Sytten 

We are currently valuating to support plain postgresql besides cockroachDB. This because multiple customers are interested in using the managed postgresql offerings.

I think we should be able to given an update to this in the next 2-4 weeks.

**Comment by Sytten** [comment]
Nice! Let me know if you need beta testers, willing to give you good feedback since I want to implement our integration with v2 as soon as possible 😀

**Comment by adlerhurst** [comment]
Support for mysql was requested here: https://github.com/zitadel/zitadel/issues/3600#issuecomment-1161246978

**Comment by KaKi87** [comment]
@hifabienne Considering you set priority to low, I'm guessing this [this statement](https://github.com/zitadel/zitadel/issues/3600#issuecomment-1168324557) from two months ago is no longer applicable ?
> Probably we will have a look at this in 1-2 months.

Thanks

**Comment by fforootd** [comment]
At the moment we are doing the finishing touches for Postgres which should be ready really soon to be released.

After this we will consider what additional databases might make sense.

**Comment by KaKi87** [comment]
Oh, very nice, thanks !

**Comment by fforootd** [comment]
What DB are you interested in?

**Comment by KaKi87** [comment]
Postgres is fine, I'll try it as soon as it's released.
Although MySQL is my primary database.

**Comment by adlerhurst** [comment]
hi @Sytten and @KaKi87 

We have just released postgresql support of ZITADEL. If you want to check it out you can find the release notes here: https://github.com/zitadel/zitadel/releases/tag/v2.3.0

If you have any feedback we are happy to answer :-)

**Comment by KaKi87** [comment]
Since PostgreSQL support of Zitadel is limited to the latest major version, I cannot use it, therefore I look forward to MySQL support. Thanks

**Comment by fforootd** [comment]
> Since PostgreSQL support of Zitadel is limited to the latest major version, I cannot use it, therefore I look forward to MySQL support. Thanks

I think it will be a while until we support mysql. Can you shine some light why it is a problem to use Postgresql 14?

**Comment by adlerhurst** [comment]
> > Since PostgreSQL support of Zitadel is limited to the latest major version, I cannot use it, therefore I look forward to MySQL support. Thanks
> 
> I think it will be a while until we support mysql. Can you shine some light why it is a problem to use Postgresql 14?

question was answered here: https://discord.com/channels/927474939156643850/1014503578439794688/1015302346449883176

**Comment by helmut72** [comment]
> If you have any feedback we are happy to answer :-)

I'm not the ones, but give feedback :-) Also wanted to switch to Postgres after seeing Cockroach RAM usage. In Keycloak the app is the memory hog, in Zitadel it's Cockroach.

This works:

```
ZITADEL_DATABASE_COCKROACH_HOST=cockroach
```

... and this doesn't work:

```
ZITADEL_DATABASE_POSTGRES_HOST=postgres
ZITADEL_DATABASE_POSTGRES_HOST=5432
ZITADEL_DATABASE_POSTGRES_DATABASE=zitadel
ZITADEL_DATABASE_POSTGRES_USER_USERNAME=zitadel
ZITADEL_DATABASE_POSTGRES_USER_PASSWORD=password
```

postgres is the container name of Postgres.

**Comment by adlerhurst** [comment]
hi @helmut72 

I think there is some configuration missing. If you use postgres you have to set all the fields for the connection. The defaults of zitadel are only configured for cockroachdb:

here is an example for disabled ssl on postgres:
```
ZITADEL_DATABASE_POSTGRES_HOST=postgres
ZITADEL_DATABASE_POSTGRES_PORT=5432
ZITADEL_DATABASE_POSTGRES_DATABASE=zitadel

ZITADEL_DATABASE_POSTGRES_USER_USERNAME=zitadel
ZITADEL_DATABASE_POSTGRES_USER_PASSWORD=password
ZITADEL_DATABASE_POSTGRES_USER_SSL_MODE=disable
```

if you init the database also with zitadel you have to set the admin user additionaly:
```
ZITADEL_DATABASE_POSTGRES_ADMIN_USERNAME=root
ZITADEL_DATABASE_POSTGRES_ADMIN_PASSWORD=root
ZITADEL_DATABASE_POSTGRES_ADMIN_SSL_MODE=disable
```

**Comment by helmut72** [comment]
@adlerhurst 
Thanks, have added SSL Mode and tried with disable, disabled, false and 0. But no luck. It still wants to connect to a Cockroach db. Which is unset:

```
level=fatal msg="unable to initialize the database" caller="/home/runner/work/zitadel/zitadel/cmd/initialise/init.go:47" error="dial tcp 127.0.0.1:26257: connect: connection refused"
```

> We are currently working on ram usage of crdb.

I hope Postgres doesn't take this much RAM too... ;) Nevertheless I prefer Postgres for my home setup.

**Comment by helmut72** [comment]
Update. Now it works. First I've pulled latest `:stable` release. My image was from end of last week. Still got an error:

```
level=fatal msg="unable to initialize the database" caller="/home/runner/work/zitadel/zitadel/cmd/initialise/init.go:65" error="ID=DATAB-0pIWD Message=Errors.Database.Connection.Failed Parent=(cannot parse `host=postgres port=5432 user= application_name=zitadel sslmode= sslrootcert=`: failed to configure TLS (sslmode is invalid))"
```

Then I've also set the Postgres admin variables:

```
ZITADEL_DATABASE_POSTGRES_ADMIN_USERNAME=zitadel
ZITADEL_DATABASE_POSTGRES_ADMIN_PASSWORD=zitadel
ZITADEL_DATABASE_POSTGRES_ADMIN_SSL_MODE=disable
```

But even then it doesn't worked:

```
level=fatal msg="unable to initialize the database" caller="/home/runner/work/zitadel/zitadel/cmd/initialise/init.go:65" error="ID=DATAB-0pIWD Message=Errors.Database.Connection.Failed Parent=(failed to connect to `host=postgres user=db_user database=`: server error (FATAL: database \"db_user\" does not exist (SQLSTATE 3D000)))"
```


Look at the database name. Same as Postgres user variable. But I have set a different db name. My examples here are other names for security reason.

Then I have tried with `zitadel` everywhere. Adminuser, Username, DB-Name and Password. Then it worked. It looks like the ENV parser need a fix with ZITADEL_DATABASE_POSTGRES_DATABASE option?

All at all it would be cool if the ENV parser use default values if any option is unset. I have this experience from other projects.

> I hope Postgres doesn't take this much RAM too... ;) 

Can answer my own question. Postgres use 38MB. For now. That's great. :)

**Comment by hifabienne** [comment]
@adlerhurst @fforootd Should we close this issue for now? We have added Postgres and I think for now we do not have the intention to add any other? Or  how do you see it?

**Comment by KaKi87** [comment]
MySQL ?

**Comment by adlerhurst** [comment]
I would leave it open that we can track the state and put it at the end of the backlog.

For now we focus to improve the currently implemented sql dialects (cockroach, postgresql). Additional dialects add an additional layer of complexity for each change and maintenance overhead. To add more dialects there must be made ground work to simplify the addition of dialects.

**Comment by musicformellons** [comment]
I could use [yugabytedb](https://www.yugabyte.com/) support. Based on postgresql, so should be 'relatively easy'.

**Comment by KaKi87** [comment]
Why not.

Although, I think MySQL is the most urgent one, as it's the most-used database.

**Comment by musicformellons** [comment]
That's right:
Black = mysql
Orange = postgres
Yellow = cockroachdb
Green = yugabytedb

<img width="498" alt="Screenshot 2023-02-01 at 15 20 17" src="https://user-images.githubusercontent.com/10334123/216069200-fe618990-367f-4730-aba1-d7cab08944ec.png">

**Comment by Wulfheart** [comment]
> I would leave it open for tracking the state and put it at the end of the backlog.
> 
>  For now we focus to improve the currently implemented sql dialects (cockroach, postgresql). Additional dialects add an additional layer of complexity for each change and maintenance overhead. To add more dialects there must be made ground work to simplify the addition of dialects.

So maybe use an ORM like sqlboiler? But that is only a shot as I am not familiar with the code. Just interested on your opinion on this.

**Comment by cr1xu5** [comment]
SQLite support would be great for compact setups on the edge, but also for testing locally.

Together with solutions like [LiteFS](https://fly.io/docs/litefs/) SQLite is a great alternative for the "regular" databases.

**Comment by gandaldf** [comment]
I don't know if anyone need this, but I was able to run Zitadel with Postgres using this docker-compose.yaml:
```
version: '3.8'

services:
  zitadel:
    restart: 'always'
    networks:
      - 'zitadel'
    image: 'ghcr.io/zitadel/zitadel:stable'
    command: 'start-from-init --masterkey "MasterkeyNeedsToHave32Characters" --tlsMode disabled'
    environment:
      - 'ZITADEL_EXTERNALSECURE=false'
      - 'ZITADEL_DATABASE_POSTGRES_HOST=pgdb'
      - 'ZITADEL_DATABASE_POSTGRES_PORT=5432'
      - 'ZITADEL_DATABASE_POSTGRES_DATABASE=zitadeldb'
      - 'ZITADEL_DATABASE_POSTGRES_ADMIN_USERNAME=zitadeladmin'
      - 'ZITADEL_DATABASE_POSTGRES_ADMIN_PASSWORD=zitadeladmin'
      - 'ZITADEL_DATABASE_POSTGRES_ADMIN_SSL_MODE=disable'
      - 'ZITADEL_DATABASE_POSTGRES_USER_USERNAME=zitadeluser'
      - 'ZITADEL_DATABASE_POSTGRES_USER_PASSWORD=zitadeluser'
      - 'ZITADEL_DATABASE_POSTGRES_USER_SSL_MODE=disable'
    depends_on:
      pgdb:
        condition: 'service_healthy'
    ports:
      - '8080:8080'

  pgdb:
    restart: 'always'
    networks:
      - 'zitadel'
    image: 'postgres:16beta1'
    environment:
      - POSTGRES_USER=zitadeladmin
      - POSTGRES_PASSWORD=zitadeladmin
      - POSTGRES_DB=zitadeldb
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: '10s'
      timeout: '30s'
      retries: 5
      start_period: '20s'
    ports:
      - '9090:8080'
      - '5432:5432'

networks:
  zitadel:
```

**Comment by unique-dominik** [comment]
I'd like to also leave a mark here as I stumbled across this issue and have a different viewpoint onto the problem as well.

Our company operates within the Financial and Insurance Sector. This sectors comes with some _limitations_ that do limit a persons choice of a database drastically (and is probably the reason why you cleverly packaged it into Enterprise 🙈 ).

- They do not especially like self-hosted databases
- They nearly all use Microsoft (few on premise, more and more in the cloud, private or public)
- They audit you/us/the database and the architecture around it
  + They like backup and restore
  + They like if its not self-made but from Microsoft and the kind
  + They like if its automated
  + They like encryption in transit and at rest
  + They like desaster recovery

Now, certainly there are services where you get Cockroach as a Service, but also this generally FIS customers do not really like, especially when the companies have 🇺🇸 -background (🇨🇭 bancs and 🇺🇸 have a special relation).

In the end, we are nearly forced, pushed / deeply motivated to take Azure-hosted databases as FIS customers trust Microsoft just more than any other startup/company that cooks their own soup.

This mindset, that large customers (and large customers bring large money) bring their own database (Cosmos, MySQL, Maria, Postgres or whatever Microsoft hosts) should be also kept in mind when evaluating this issue 🤔 

Sadly, that is also why most of them and us can and won't use Zitadel Cloud. They want it in their own Azure Tenant.

**Visionary thought:** Given that you are 🇨🇭 based and maybe one day want to bark up the FIS sector as well (just dreaming) you would kind of start to fall into the [Landing Zone Concept](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/landing-zone/) like we do and there, it would be (so I assume) _highly_ encourage that you use one Microsoft DB's 🚀

**Comment by KaKi87** [comment]
That's a closed-mindedness issue, i.e. invalid from my point of view.

**Comment by fforootd** [comment]
> I'd like to also leave a mark here as I stumbled across this issue and have a different viewpoint onto the problem as well.
> 
> 
> 
> Our company operates within the Financial and Insurance Sector. This sectors comes with some _limitations_ that do limit a persons choice of a database drastically (and is probably the reason why you cleverly packaged it into Enterprise 🙈 ).
> 
> 
> 
> - They do not especially like self-hosted databases
> 
> - They nearly all use Microsoft (few on premise, more and more in the cloud, private or public)
> 
> - They audit you/us/the database and the architecture around it
> 
>   + They like backup and restore
> 
>   + They like if its not self-made but from Microsoft and the kind
> 
>   + They like if its automated
> 
>   + They like encryption in transit and at rest
> 
>   + They like desaster recovery
> 
> 
> 
> Now, certainly there are services where you get Cockroach as a Service, but also this generally FIS customers do not really like, especially when the companies have 🇺🇸 -background (🇨🇭 bancs and 🇺🇸 have a special relation).
> 
> 
> 
> In the end, we are nearly forced, pushed / deeply motivated to take Azure-hosted databases as FIS customers trust Microsoft just more than any other startup/company that cooks their own soup.
> 
> 
> 
> This mindset, that large customers (and large customers bring large money) bring their own database (Cosmos, MySQL, Maria, Postgres or whatever Microsoft hosts) should be also kept in mind when evaluating this issue 🤔 
> 
> 
> 
> Sadly, that is also why most of them and us can and won't use Zitadel Cloud. They want it in their own Azure Tenant.
> 
> 
> 
> **Visionary thought:** Given that you are 🇨🇭 based and maybe one day want to bark up the FIS sector as well (just dreaming) you would kind of start to fall into the [Landing Zone Concept](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/landing-zone/) like we do and there, it would be (so I assume) _highly_ encourage that you use one Microsoft DB's 🚀 

Thank you for this insightful write up 🙏.

Allow me a question, would it not be sufficient for most to run zitadel inside their azure tenant and use one of the managed Postgres offerings from azure? Or do these customers want a managed zitadel in their own tenant? 

Because this is what we see people doing instead of using our cloud service.

(Postgres works in the OSS version already, just yesterday I deployed a zitadel with Google cloud SQL) 

As a side note CRDB soon will offer Azure as well, but I do not know the details 😅.

**Comment by fforootd** [comment]
> That's a closed-mindedness issue, i.e. invalid from my point of view.

I am not sure I am getting this.

What is closed-minded?

**Comment by KaKi87** [comment]
> We do not want that, our customers want that.

I know, but you should be convincing them to open themselves to what the rest of the world can offer, despite the additional efforts required, and eventually help them doing these.

Otherwise, we'll never stop GAFAM from ruling us.

**Comment by fforootd** [comment]
> 🤝
> 
> > Allow me a question, would it not be sufficient for most to run zitadel inside their azure tenant and use one of the managed Postgres offerings from azure? Or do these customers want a managed zitadel in their own tenant?
> 
> We see both. We have one customer asking for a fully managed service (where we actually struggle to find a good service-level agreement kind of because Privileged and Conditional Access to now _their_ tenant). But we also have one customer that wants our solution deployed over a LZ but then manages it mostly themselves. We are also still finding our way through Microsoft and their LZ/Marketplace visions and papers but in the end the customers really just care about: Where is the data, who sees/processes it when and for how long and are they in charge of all the encryption keys or not.
> 
> > Because this is what we see people doing instead of using our cloud service.
> 
> Yeah, like us.
> 
> > Because this is what we see people doing instead of using our cloud service.
> 
> I am right now running terraform code to provision an Azure Flexible PSQL to do exactly that 🚀
> 
> > As a side note CRDB soon will offer Azure as well, but I do not know the details 😅
> 
> This would be very interesting if they allow customer-managed encryption keys 🤔
> 
> @KaKi87 We do not want that, our customers want that. And they are regulated by authorities and the law so you get edged into a 🗳️ in these situations. Of course you can self-host but then you have to answer every audit why exactly you chose GPG as your favourite key and who exactly can access the key you made on your local machine, when did you rotate it last (show the evidence you did) etc. etc. etc.
> 
> **Update:** You can have a look at this document from an expert in 🇨🇭 about that [Does a cloud solution meet the requirements of a Swiss bank?](https://www.rosenthal.ch/downloads/VISCHER-Swiss-Banks-Cloud.pdf) to get an impression of the situation.

Yeah there will always be a need for regulated customers to take more control over their infra and data.

We had the idea once that we could deploy zitadel into customers infra through our customer portal (so basically we deploy to your tenant and provision all the things we need). However we postponed that since it not an easy task to achieve.
An alternative could be the marketplaces, however that is not easy/cheap either 😁 

This is why we currently think having postgres and good support for k8s/knative is a great approach.

**Comment by unique-dominik** [comment]
> This is why we currently think having postgres and good support for k8s/knative is a great approach.

Sure! I just left my thoughts as notes in case you revisit the issue one day 🚀 And since it works since yesterday with AKS Flex PSQL we can cover most of the audited world ourselves already 💪

**Comment by moay** [comment]
> Thank you for this insightful write up 🙏.
> 
> Allow me a question, would it not be sufficient for most to run zitadel inside their azure tenant and use one of the managed Postgres offerings from azure? Or do these customers want a managed zitadel in their own tenant?

Short addition here @fforootd: We are one of those customers and currently cannot switch to Zitadel because of the missing MySQL support. I'd love to, but our existing infrastructure is all MySQL, including backups, security agreements, SLAs, legal stuff, management/escalation agreements and so on. Our team is too small/busy to get into longer conversations with the legal department/compliance department/etc  why we can't use existing resources, who will maintain Postgres in case of disaster, etc..

The hurdles to take might be out of control of the dev department and sometimes there is a lot of enterprise overhead in bigger structures. As I'd assume supporting MySQL should be just about adding some existing driver, I'd be really happy to see this feature and start switching over to Zitadel.

**Comment by holotrack** [comment]
Any news on MySQL/TiDB support?

**Comment by hifabienne** [comment]
> Any news on MySQL/TiDB support?

At the moment we do not have any news regarding support of mysql or any other databases. 
A new database always brings additional complexity into the product, as not every DB works exactly the same. We can already see that with cockcroach and PostgreSQL.
Right now, our focus is on improving the current integrations, not adding more. However, this might change in the future. So we still hold the issue open to get all your feedback and opinion.

**Comment by holotrack** [comment]
> At the moment we do not have any news regarding support of mysql or any other databases. A new database always brings additional complexity into the product, as not every DB works exactly the same. We can already see that with cockcroach and PostgreSQL. Right now, our focus is on improving the current integrations, not adding more. However, this might change in the future. So we still hold the issue open to get all your feedback and opinion.

Maybe im missing something but why not to use some ORM?

**Comment by BenjaminGK** [comment]
We are a big enterprise company with several sister companies and only uses MYSQL, any plans on supporting this anytime soon as it is the only reason preventing us from using your solution?

**Comment by KaKi87** [comment]
> We are a big enterprise company with several sister companies and only uses MYSQL

That doesn't make any sense.

It's like saying "we only use SOAP APIs powered by Java running on Ubuntu Server" : sure it's common but still a ridiculous meaningless limitation. 🤷

**Comment by jnischler** [comment]
I think that cloud spanner would also be a good choice.
Of course especially for the google cloud users ;)

We are currently using spanner to store permission on a global scale.
We would like to do the same with our user identities ...

Spanner should behave quite similar to Cockroach so maybe porting isn't that hard compared to other types of databases.

Thanks

**Comment by fforootd** [comment]
> We are a big enterprise company with several sister companies and only uses MYSQL, any plans on supporting this anytime soon as it is the only reason preventing us from using your solution?

Nice to hear from you.

At the moment we have not planned mysql compatibility but I would encourage you to use a postgres/cockroachdb since we have a lot of experience with these storage layers. 

The main reason to (not yet) support other storage layers is that we currently lack the engineering capacity to do this in a high quality 😁 I hope you understand this.

**Comment by fforootd** [comment]
> I think that cloud spanner would also be a good choice. Of course especially for the google cloud users ;)
> 
> We are currently using spanner to store permission on a global scale. We would like to do the same with our user identities ...
> 
> Spanner should behave quite similar to Cockroach so maybe porting isn't that hard compared to other types of databases.
> 
> Thanks

@adlerhurst and I where playing around with the postgres driver of spanner in the past but never finished it 😁 

But certainly a postgres spanner optimization would be an interesting thing.

**Comment by iFrozenPhoenix** [comment]
Indeed SQLite support or an other embedded database would be really great for smaller deployments.

**Comment by ext4xfs** [comment]
Agree, if you were to implement sqlite support, I think you'd get a lot of traction.  A lightweight db would be much better for CI/CD for infrastructure.

**Comment by iFrozenPhoenix** [comment]
@ext4xfs if I get 15 thumbs up I do it if they agree that they would accept a PR 😊

**Comment by michelkaeser** [comment]
> Indeed SQLite support or an other embedded database would be really great for smaller deployments.

+1 on this, even though my primary use-case would be simplified testing/setup in CI/CD.

**Comment by kaushalyap** [comment]
@fforootd Any plans to support SQLite in the near future?

**Comment by iFrozenPhoenix** [comment]
@fforootd would you accept a community PR for SQLite support?

**Comment by fforootd** [comment]
> @fforootd would you accept a community PR for SQLite support?

I can clarify with them team, what I want to make clear though this is not something we likely tend to support as part of our offering 😁 

@adlerhurst what is your take, do you think it is even feasible?

**Comment by adlerhurst** [comment]
We are currently working on database changes which could influence an implementation of another database quite heavily therefore i would wait with that feature until we gained some clarity there. 

What i am concerned about is the maintenance overhead after the implementation. It's already time consuming maintaining crdb and postgres.

**Comment by SergeAx** [comment]
As an avid self-hoster, I would vouch for SQLite and even consider working on community PR. Postgres is a popular database but hard to maintain, as you mentioned in the docs. But getting crdb into self-hosted stack is even more challenging for those without prior knowledge.

On the other hand, with today's hardware, SQLite is more than enough for enthusiast projects and requires no maintenance at all. It even moves toward native horizontal scaling [1], so the possible higher availability scenario in 1-2 years can be just WAL-replicated SQLite.

[1]: https://sqlite.org/draft/rsync.html

**Comment by Wulfheart** [comment]
> so the possible higher availability scenario in 1-2 years can be just WAL-replicated SQLite.

A REPLICA is readonly. I don’t think that it works so easy.

**Comment by Mark3xtrm** [comment]
Hi, guys! So no update on this?

I think the implementation of a lighter db like sqlite would make Zitadel an even more attractive option for home lab and smart home users, or for those who wanted a more contained choice than CRDB or PG.

I understand that it is not easy already to maintain two DBs, but I hope you can consider the idea.

Thank you for your valuable time, in any case

**Comment by iFrozenPhoenix** [comment]
@adlerhurst @fforootd Could you inform us if the changes were made and if you could clarify if a community PR for SQLite would be accepted?

**Comment by adlion** [comment]
+1 for SQLite or libsql

**Comment by wildtang** [comment]
> +1 for SQLite or libsql

We are a small team, has deployed some open source and our own service.
Someone of them support LDAP, someone support OAuth, etc.
so we hope has one light weight account system can support auth methods as many.
Because we are a small team, so we thing a light weight SQLite database is enough for storing the accounts data.

**Comment by yimiaoxiehou** [comment]
i think embed postgresql( like https://github.com/fergusstrange/embedded-postgres) into ztialdel is a good idea, we just want to make system usage less and deploy convenient，sqlite isn‘t the only way

**Comment by yimiaoxiehou** [comment]
i think embed postgresql( like https://github.com/fergusstrange/embedded-postgres) into ztialdel is a good idea, we just want to make system usage less and deploy convenient，sqlite isn‘t the only way

**Comment by fforootd** [comment]
> i think embed postgresql( like https://github.com/fergusstrange/embedded-postgres) into ztialdel is a good idea, we just want to make system usage less and deploy convenient，sqlite isn‘t the only way

That is an interesting project. What is your use-case for that? Test envs or simple non scalalble deployments?

**Comment by divStar** [comment]
SQLite would be very welcome indeed, because it would allow deployments on machines with less RAM.

**Comment by geogeim** [comment]
I want to run Zitadel onto a raspberry pi that's currently handling homeassistant and a few other things. It would be really nice if an SQLite option for storage would be available

**Comment by Stoorx** [comment]
+1 to MySQL/MariaDB. PostgreSQL HA clustering is pita.

**Comment by pomac303** [comment]
Agreed, galera is easy to setup, i have found nothing for postgresql that is still alive... :/

**Comment by guenhter** [comment]
Also SQLite would be very convenient for testing purposes...
